### PR TITLE
add Azure machine pool rolling upgrades with MaxSurge, MaxUnavailable and DeletePolicy

### LINKED ISSUE #1361: MachinePool has more worker nodes than the desired value
/kind bug

**What steps did you take and what happened:**
[A clear and concise description of what the bug is.]

https://prow.k8s.io/view/gs/kubernetes-jenkins/logs/capz-azure-file-machinepool-1-19/1388742406299455488

```
Waiting for 1 control plane machine(s) and 2 worker machine(s) to become Ready
node/capz-h8u4r3-control-plane-5m8mq condition met
node/capz-h8u4r3-mp-0000000 condition met
node/capz-h8u4r3-mp-0000002 condition met
NAME                              STATUS     ROLES    AGE     VERSION                           INTERNAL-IP   EXTERNAL-IP   OS-IMAGE             KERNEL-VERSION     CONTAINER-RUNTIME
capz-h8u4r3-control-plane-5m8mq   Ready      master   3m59s   v1.19.11-rc.0.23+3ed2eb74c47e2f   10.0.0.4      <none>        Ubuntu 18.04.5 LTS   5.3.0-1034-azure   containerd://1.3.4
capz-h8u4r3-mp-0000000            Ready      <none>   37s     v1.19.11-rc.0.23+3ed2eb74c47e2f   10.1.0.4      <none>        Ubuntu 18.04.5 LTS   5.3.0-1034-azure   containerd://1.3.4
capz-h8u4r3-mp-0000001            NotReady   <none>   8s      v1.19.11-rc.0.23+3ed2eb74c47e2f   10.1.0.5      <none>        Ubuntu 18.04.5 LTS   5.3.0-1034-azure   containerd://1.3.4
capz-h8u4r3-mp-0000002            Ready      <none>   20s     v1.19.11-rc.0.23+3ed2eb74c47e2f   10.1.0.6      <none>        Ubuntu 18.04.5 LTS   5.3.0-1034-azure   containerd://1.3.4
```

Eventually, the `NodeReady` node was removed.

**What did you expect to happen:**

Have no more than the desired MachinePool replicas at any given moment

**Anything else you would like to add:**
[Miscellaneous information that will assist in solving the issue.]

**Environment:**

- cluster-api-provider-azure version: master
- Kubernetes version: (use `kubectl version`): 
- OS (e.g. from `/etc/os-release`):

### PR DESCRIPTION
**Pull request by devigned**
**What type of PR is this?**
/kind feature

**What this PR does / why we need it**:
This PR introduces `AzureMachinePoolMachine` resource to replace `AzureMachinePool.Status.Instances` in accordance to [Azure Machine Pool Machines Proposal](https://github.com/kubernetes-sigs/cluster-api-provider-azure/blob/adaf9c6d4b03bbe42174ad2dd1a7c0fba66a56dd/docs/proposals/20210222-azure-machinepool-machine.md).

This also introduces the machine pool rolling upgrade strategy which provides, MaxSurge, MaxUnavailable, and DeletePolicy modeled after machine deployments.

Functional changes:
- Ability to delete individual `AzureMachinePoolMachines`
- MaxSurge to surge machine pools upon upgrade of the machine pool template
- MaxUnavailable to ensure a minimum number of machines are available at any time
- DeletePolicy to enable preference on which nodes to delete, Oldest, Newest, Random
- Fully async reconciliation of `AzureMachinePool` resources

Non-goals of the PR (follow up PRs):
- implement cordon and drain per Azure machine pool machine
- implement instance update rather than delete / re-create

**Which issue(s) this PR fixes** *(optional, in `fixes #<issue number>(, fixes #<issue_number>, ...)` format, will close the issue(s) when PR gets merged)*:
related #1105
fixes #1361

**Special notes for your reviewer**:

_Please confirm that if this PR changes any image versions, then that's the sole change this PR makes._

**TODOs**:

- [x] squashed commits
- [ ] includes documentation
- [x] adds unit tests

**Release note**:

```release-note
add Azure machine pool rolling upgrades with MaxSurge, MaxUnavailable and DeletePolicy. 
```

### CODE REVIEW

File: exp/api/v1alpha4/azuremachinepool_types.go [hunk 1]
```diff
@@ -111,6 +129,95 @@ type (
...
+	// inform the controller that if the parent MachinePool.Spec.Template.Spec.Version field is updated, this image
+	// will be updated with the corresponding default image. If Defaulted is set to false, the controller will not
+	// update the image reference when the MachinePool.Spec.Template.Spec.Version changes.
+	AzureDefaultingImage struct {
+		infrav1.Image `json:",omitempty"`
+
+		// Defaulted informs the controller that the image reference was defaulted so that it can be updated by changes
+		// to the MachinePool.Spec.Template.Spec.Version field by default.
+		// +optional
+		Defaulted bool `json:"defaulted,omitempty"`
```
Comment (CecileRobertMichon) [inline exp/api/v1alpha4/azuremachinepool_types.go:220]: What's the UX for this? Adding this as a spec field means it can technically be set by the user. 

Is it essential to achieving the goal of this PR? If not, would you consider taking this change out and separating it to be is own PR? This would allow us to align the other uses of Image and to carefully consider the UX and back compat implications of a change like this instead of having it be a small part in this XXL PR.

**Review by CecileRobertMichon — COMMENTED** [review]
I started reviewing but I'm having a hard time following all the different changes... It would greatly help review if you were able to split out any unrelated improvements to other PRs and structure the changes in this PR in multiple commits.

File: azure/scope/machinepool.go [hunk 2]
```diff
@@ -58,12 +57,14 @@ type (
...
-		patchHelper      *patch.Helper
-		MachinePool      *capiv1exp.MachinePool
-		AzureMachinePool *infrav1exp.AzureMachinePool
 		azure.ClusterScoper
+		logr.Logger
+		AzureMachinePool   *infrav1exp.AzureMachinePool
+		MachinePool        *capiv1exp.MachinePool
+		client             client.Client
+		patchHelper        *patch.Helper
+		vmssState          *infrav1exp.VMSS
```
[the remark this answers is not in the document]
Comment (devigned) [reply azure/scope/machinepool.go]: `ScaleSet.Reconcile` sets the state of the VMSS on the scope so that the scope can properly update the status of the `AzureMachinePool`. I don't think the information would be useful to set directly on the `AzureMachinePool`, but it does lead to more concise information being set on the `AzureMachinePool`, like conditions, etc.

File: azure/scope/machinepool.go [hunk 3]
```diff
@@ -158,61 +163,217 @@ func (m *MachinePoolScope) ProvisioningState() infrav1.ProvisioningState {
...
+	if m.vmssState != nil {
+		for _, machine := range m.vmssState.Instances {
+			if !machine.LatestModelApplied {
+				return true
+			}
+		}
+	}
+
+	var (
+		desiredMatchesActual = len(m.vmssState.Instances) == int(m.DesiredReplicas())
```
[the remark this answers is not in the document]
Comment (devigned) [reply azure/scope/machinepool.go:180]: good point

File: exp/api/v1alpha4/azuremachinepool_types.go (hunk 1 above)
Comment (devigned) [reply exp/api/v1alpha4/azuremachinepool_types.go:220]: The UI is that the Defaulted can be set or unset by the user. If Default == true, then the default image version will be used for that version of K8s. If false, then a specific image reference will be needed.

This will be essential for passing the upgrade test as it is today where a machine pool's version is set and the machine pool is expected to rollout a new version of K8s.

File: test/e2e/helpers.go [hunk 4]
```diff
@@ -327,28 +331,98 @@ func getClusterSSHInfo(ctx context.Context, c client.Client, namespace, name str
...
+		}
+
+		nodes = append(nodes, node)
+	}
+
+	return nodes, nil
+}
+
+func getWorkloadClient(ctx context.Context, c client.Client, namespace, clusterName string) (client.Client, error) {
+	ctx, span := tele.Tracer().Start(ctx, "scope.MachinePoolMachineScope.getWorkloadClient")
```
Comment (CecileRobertMichon) [inline test/e2e/helpers.go:401]: why scope.MachinePoolMachineScope ?

File: test/e2e/helpers.go [hunk 5]
```diff
@@ -327,28 +331,98 @@ func getClusterSSHInfo(ctx context.Context, c client.Client, namespace, name str
...
+			continue
+		}
+
+		nodes = append(nodes, node)
+	}
+
+	return nodes, nil
+}
+
+func getWorkloadClient(ctx context.Context, c client.Client, namespace, clusterName string) (client.Client, error) {
```
Comment (CecileRobertMichon) [inline test/e2e/helpers.go:400]: is there no helper in the test framework that we can use for this?

File: api/v1alpha4/conditions_consts.go [hunk 6]
```diff
@@ -63,3 +63,25 @@ const (
 	// ScaleSetProvisionFailedReason used for failures during scale set provisioning.
 	ScaleSetProvisionFailedReason = "ScaleSetProvisionFailed"
 )
+
+// AzureMachinePool Conditions and Reasons
+const (
+	// PoolRunningCondition reports on current status of the Azure scale set.
+	PoolRunningCondition clusterv1.ConditionType = "PoolRunning"
```
Comment (CecileRobertMichon) [inline api/v1alpha4/conditions_consts.go:70]: what's the difference between `ScaleSetRunningCondition` above and `PoolRunningCondition`? They have the same description

File: azure/scope/machinepool.go [hunk 7]
```diff
@@ -158,61 +163,216 @@ func (m *MachinePoolScope) ProvisioningState() infrav1.ProvisioningState {
...
-	for i, instance := range instances {
-		instanceStatuses[i] = &infrav1exp.AzureMachinePoolInstanceStatus{
-			ProviderID:        azure.ProviderIDPrefix + instance.ID,
-			InstanceID:        instance.InstanceID,
-			InstanceName:      instance.Name,
-			ProvisioningState: &instance.State,
+	providerIDs := make([]string, len(machines))
+	for i, machine := range machines {
+		if machine.Status.Ready {
+			readyReplicas++
```
Comment (CecileRobertMichon) [inline azure/scope/machinepool.go:221]: lots of newly added logic in here, we really need to start adding unit test to these files

File: azure/types.go [hunk 8]
```diff
@@ -206,3 +210,71 @@ type VMSSExtensionSpec struct {
...
+	instancesByProviderID := make(map[string]VMSSVM, len(vmss.Instances))
+	for _, instance := range vmss.Instances {
+		instancesByProviderID[instance.ProviderID()] = instance
+	}
+
+	return instancesByProviderID
+}
+
+// ProviderID returns the K8s provider ID for the VMSS instance
+func (vm VMSSVM) ProviderID() string {
```
Comment (CecileRobertMichon) [inline azure/types.go:261]: does this stuff really belong in here?

File: azure/types.go [hunk 9]
```diff
@@ -206,3 +210,71 @@ type VMSSExtensionSpec struct {
...
+	return fmt.Sprintf("azure://%s", vm.ID)
```
Comment (CecileRobertMichon) [inline azure/types.go:262]: please use the const for azure://

**Review by CecileRobertMichon — COMMENTED** [review]
still working through reviewing all the files... thanks for splitting up commits

File: azure/scope/machinepool.go (hunk 7 above)
Comment (devigned) [reply azure/scope/machinepool.go:221]: Couldn't agree more. I was hesitant to add too much more to this PR.

File: azure/types.go (hunk 8 above)
Comment (devigned) [reply azure/types.go:261]: I'm happy to put them where ever. They definitely didn't belong in the `./exp/api` where they were before. 

I do think they belong here as they are basically the spec structs which have yet to be refactored for the exp controllers.

File: api/v1alpha4/conditions_consts.go (hunk 6 above)
Comment (devigned) [reply api/v1alpha4/conditions_consts.go:70]: I think this was a merge resolution issue. Fixing.

File: test/e2e/helpers.go (hunk 4 above)
Comment (devigned) [reply test/e2e/helpers.go:401]: deleting...

File: test/e2e/helpers.go (hunk 5 above)
Comment (devigned) [reply test/e2e/helpers.go:400]: Good call. Found it via the BootstrapClusterProxy.

**Review by nader-ziada — COMMENTED** [review]
just some small comments, mostly cause it hard to follow :) I think I get the big picture, but will need to go through the `scope` package again

**Comment by devigned** [comment]
/hold
for additional test coverage

**Comment by nader-ziada** [comment]
went through it again and mI think its mostly good; the more unit tests will help make the `scope` changes easier to understand and more solid for sure

File: exp/controllers/azuremachinepoolmachine_controller.go [hunk 10]
```diff
@@ -0,0 +1,389 @@
...
+
+	return reconcile.Result{}, nil
+}
+
+func newAzureMachinePoolMachineReconciler(scope *scope.MachinePoolMachineScope) azure.Reconciler {
+	return &azureMachinePoolMachineReconciler{
+		Scope:              scope,
+		scalesetVMsService: scalesetvms.NewService(scope),
+	}
+}
```
[the remark this answers is not in the document]
Comment (devigned) [reply exp/controllers/azuremachinepoolmachine_controller.go:351]: It is now used in the tests.

File: exp/controllers/azuremachinepoolmachine_controller.go [hunk 11]
```diff
@@ -0,0 +1,389 @@
...
+
+// NewAzureMachinePoolMachineController creates a new AzureMachinePoolMachineController to handle updates to Azure Machine Pool Machines
+func NewAzureMachinePoolMachineController(c client.Client, log logr.Logger, recorder record.EventRecorder, reconcileTimeout time.Duration, watchFilterValue string) *AzureMachinePoolMachineController {
+	return &AzureMachinePoolMachineController{
+		Client:            c,
+		Log:               log,
+		Recorder:          recorder,
+		ReconcileTimeout:  reconcileTimeout,
+		WatchFilterValue:  watchFilterValue,
+		reconcilerFactory: newAzureMachinePoolMachineReconciler,
```
[the remark this answers is not in the document]
Comment (devigned) [reply exp/controllers/azuremachinepoolmachine_controller.go:82]: This is now used in the tests.

File: exp/controllers/azuremachinepoolmachine_controller.go (hunk 10 above)
[the remark this answers is not in the document]
Comment (devigned) [reply exp/controllers/azuremachinepoolmachine_controller.go:351]: This reconciler is used to tie together the underlying Azure service reconciles with updating the scope.

File: exp/api/v1alpha4/azuremachinepool_types.go (hunk 1 above)
Comment (devigned) [reply exp/api/v1alpha4/azuremachinepool_types.go:220]: I've added a test to better describe the behavior of this change.

**Comment by devigned** [comment]
On reconcile AzureMachinePool will save the VMImage in the Status of the AzureMachinePool and keep the Spec Image nil while defaulting to the capi offer image. If the AzureMachinePool Spec Image is set, then it will be used and stored on the Status Image.

**Comment by devigned** [comment]
@nader-ziada and @CecileRobertMichon is there anything else you'd like to see changed in this PR?

**Comment by nader-ziada** [comment]
> @nader-ziada and @CecileRobertMichon is there anything else you'd like to see changed in this PR?

lgtm

**Comment by CecileRobertMichon** [comment]
I'll take another pass by EOD

File: api/v1alpha4/types.go [hunk 12]
```diff
@@ -587,3 +587,8 @@ type AzureBastion struct {
 	// +optional
 	PublicIP PublicIPSpec `json:"publicIP,omitempty"`
 }
+
+// IsTerminalVMState returns true if the VMState is a terminal state for an Azure resource
+func IsTerminalVMState(state ProvisioningState) bool {
```
Comment (CecileRobertMichon) [inline api/v1alpha4/types.go:592]:
```suggestion
func IsTerminalProvisioningState(state ProvisioningState) bool {
```

File: api/v1alpha4/types.go [hunk 13]
```diff
@@ -587,3 +587,8 @@ type AzureBastion struct {
 	// +optional
 	PublicIP PublicIPSpec `json:"publicIP,omitempty"`
 }
+
+// IsTerminalVMState returns true if the VMState is a terminal state for an Azure resource
```
Comment (CecileRobertMichon) [inline api/v1alpha4/types.go:591]:
```suggestion
// IsTerminalProvisioningState returns true if the ProvisioningState is a terminal state for an Azure resource
```

File: azure/converters/vmss.go [hunk 14]
```diff
@@ -46,25 +45,65 @@ func SDKToVMSS(sdkvmss compute.VirtualMachineScaleSet, sdkinstances []compute.Vi
...
+// SDKImageToImage converts a SDK image reference to infrav1.Image
+func SDKImageToImage(sdkImageRef *compute.ImageReference) infrav1.Image {
+	return infrav1.Image{
+		ID: sdkImageRef.ID,
+		Marketplace: &infrav1.AzureMarketplaceImage{
+			Publisher:       to.String(sdkImageRef.Publisher),
+			Offer:           to.String(sdkImageRef.Offer),
+			SKU:             to.String(sdkImageRef.Sku),
+			Version:         to.String(sdkImageRef.Version),
+			ThirdPartyImage: false,
```
Comment (CecileRobertMichon) [inline azure/converters/vmss.go:106]: why is this hardcoded to false? I believe the way to check that is to check if their is a Plan defined on the vm

File: azure/converters/vmss.go [hunk 15]
```diff
@@ -46,25 +45,65 @@ func SDKToVMSS(sdkvmss compute.VirtualMachineScaleSet, sdkinstances []compute.Vi
...
+		instance.Name = *sdkInstance.OsProfile.ComputerName
+	}
+
+	if sdkInstance.StorageProfile != nil && sdkInstance.StorageProfile.ImageReference != nil {
+		imageRef := sdkInstance.StorageProfile.ImageReference
+		instance.Image = SDKImageToImage(imageRef)
+	}
+
+	if sdkInstance.Zones != nil && len(*sdkInstance.Zones) > 0 {
+		instance.AvailabilityZone = to.StringSlice(sdkInstance.Zones)[0]
```
Comment (CecileRobertMichon) [inline azure/converters/vmss.go:91]: why is this hardcoded to zero? Do we expect any cases where there is more than 1 zone there? If not, can you please add a code comment to indicate that?

File: azure/errors.go [hunk 16]
```diff
@@ -109,6 +109,11 @@ func (t ReconcileError) IsTerminal() bool {
 	return t.errorType == TerminalErrorType
 }
 
+// Is returns true if the target is an OperationNotDoneError
```
Comment (CecileRobertMichon) [inline azure/errors.go:112]: how is OperationNotDoneError related to this func? seems generic

File: azure/scope/machinepoolmachine.go [hunk 17]
```diff
@@ -0,0 +1,366 @@
...
+}
+
+func newWorkloadClusterProxy(c client.Client, cluster client.ObjectKey) *workloadClusterProxy {
+	return &workloadClusterProxy{
+		Client:  c,
+		Cluster: cluster,
+	}
+}
+
+func (np *workloadClusterProxy) GetNodeByObjectReference(ctx context.Context, nodeRef corev1.ObjectReference) (*corev1.Node, error) {
```
Comment (CecileRobertMichon) [inline azure/scope/machinepoolmachine.go:290]: missing comment

File: azure/scope/machinepoolmachine_test.go [hunk 18]
```diff
@@ -0,0 +1,355 @@
...
+	mock_scope "sigs.k8s.io/cluster-api-provider-azure/azure/scope/mocks"
+	infrav1 "sigs.k8s.io/cluster-api-provider-azure/exp/api/v1alpha4"
+	gomock2 "sigs.k8s.io/cluster-api-provider-azure/internal/test/matchers/gomock"
+	clusterv1 "sigs.k8s.io/cluster-api/api/v1alpha4"
+	capiv1exp "sigs.k8s.io/cluster-api/exp/api/v1alpha4"
+	"sigs.k8s.io/controller-runtime/pkg/client/fake"
+)
+
+const (
+	ProviderID = "/foo/bin/bazz"
```
Comment (CecileRobertMichon) [inline azure/scope/machinepoolmachine_test.go:41]:
```suggestion
	FakeProviderID = "/foo/bin/bazz"
```

File: exp/controllers/azuremachinepoolmachine_controller.go [hunk 19]
```diff
@@ -0,0 +1,389 @@
...
+		reconcilerFactory azureMachinePoolMachineReconcilerFactory
+	}
+
+	azureMachinePoolMachineReconciler struct {
+		Scope              *scope.MachinePoolMachineScope
+		scalesetVMsService *scalesetvms.Service
+	}
+)
+
+// NewAzureMachinePoolMachineController creates a new AzureMachinePoolMachineController to handle updates to Azure Machine Pool Machines
```
Comment (CecileRobertMichon) [inline exp/controllers/azuremachinepoolmachine_controller.go:74]:
```suggestion
// NewAzureMachinePoolMachineController creates a new AzureMachinePoolMachineController to handle updates to Azure Machine Pool Machines.
```

File: main.go [hunk 20]
```diff
@@ -295,18 +305,37 @@ func main() {
 
 	// Setup the context that's going to be used in controllers and for the manager.
 	ctx := ctrl.SetupSignalHandler()
+	registerControllers(ctx, mgr)
+	// +kubebuilder:scaffold:builder
 
-	if err = controllers.NewAzureMachineReconciler(
-		mgr.GetClient(),
-		ctrl.Log.WithName("controllers").WithName("AzureMachine"),
+	if err := mgr.AddReadyzCheck("ping", healthz.Ping); err != nil {
```
Comment (CecileRobertMichon) [inline main.go:311]: is it normal that readyz check takes the same args as healthz check?

File: pkg/coalescing/reconciler.go [hunk 21]
```diff
@@ -0,0 +1,121 @@
...
+
+	// reconciler is the caching reconciler middleware that uses the cache or
+	reconciler struct {
+		upstream reconcile.Reconciler
+		cache    ReconcileCacher
+		log      logr.Logger
+	}
+)
+
+// NewRequestCache creates a new instance of a ReconcileCache given a specified window of expiration
```
Comment (CecileRobertMichon) [inline pkg/coalescing/reconciler.go:56]:
```suggestion
// NewRequestCache creates a new instance of a ReconcileCache given a specified window of expiration.
```

do we not have the exported func linter enabled?

File: test/e2e/config/azure-dev.yaml [hunk 22]
```diff
@@ -83,15 +86,12 @@ providers:
...
-    replacements:
-    - old: "--v=0"
-      new: "--v=4"
 
 variables:
   KUBERNETES_VERSION: "${KUBERNETES_VERSION:-v1.19.7}"
   ETCD_VERSION_UPGRADE_TO: "3.4.3-0"
   COREDNS_VERSION_UPGRADE_TO: "1.6.7"
-  KUBERNETES_VERSION_UPGRADE_TO: "${KUBERNETES_VERSION_UPGRADE_TO:-v1.19.7}"
+  KUBERNETES_VERSION_UPGRADE_TO: "${KUBERNETES_VERSION_UPGRADE_TO:-v1.19.10}"
```
Comment (CecileRobertMichon) [inline test/e2e/config/azure-dev.yaml:94]:
```suggestion
  KUBERNETES_VERSION_UPGRADE_TO: "${KUBERNETES_VERSION_UPGRADE_TO:-v1.19.11}"
```

File: templates/cluster-template-machinepool-system-assigned-identity.yaml [hunk 23]
```diff
@@ -162,6 +162,12 @@ metadata:
 spec:
   identity: SystemAssigned
   location: ${AZURE_LOCATION}
+  strategy:
```
Comment (CecileRobertMichon) [inline templates/cluster-template-machinepool-system-assigned-identity.yaml:165]: are these required changes to the templates or are they illustrative / used as examples? I don't see any docs updates or additions in this PR, what parts of the PR do you think are user-facing changes that need documentation updates?

File: azure/converters/vmss.go (hunk 15 above)
Comment (devigned) [reply azure/converters/vmss.go:91]: Instance can only be in 1 zone. I believe the same thing happens for VMs. I'll add a comment.

File: templates/cluster-template-machinepool-system-assigned-identity.yaml (hunk 23 above)
Comment (devigned) [reply templates/cluster-template-machinepool-system-assigned-identity.yaml:165]: These are not required. I only added them for illustrative purposes.

The item that needs documentation is the rollout and delete policies. I'll create an issue to address rather than adding to an already enormous PR.

File: main.go (hunk 20 above)
Comment (devigned) [reply main.go:311]: I agree. I was just keeping what is already in HEAD.

https://github.com/kubernetes-sigs/cluster-api-provider-azure/blob/bd904772ec98f83631c4528260f0c55a2a6efbcf/main.go#L431-L434

Comment (devigned) [reply main.go:311]: Oddly, the same in CAPI:
https://github.com/kubernetes-sigs/cluster-api/blob/45a0abb58cf1bcd65c9b35b0f7e189ac2cae6f0a/main.go#L217-L220

File: azure/scope/machinepoolmachine.go (hunk 17 above)
Comment (devigned) [reply azure/scope/machinepoolmachine.go:291]: Fixed

File: exp/api/v1alpha4/azuremachinepool_types.go (hunk 1 above)
Comment (devigned) [reply exp/api/v1alpha4/azuremachinepool_types.go]: Replaced with status tracking after design discussion.

**Comment by CecileRobertMichon** [comment]
test failed because of quota issues in centralus: #1428

**Comment by CecileRobertMichon** [comment]
this fixes #819 and https://github.com/kubernetes-sigs/cluster-api-provider-azure/issues/1361, correct?

**Comment by CecileRobertMichon** [comment]
/lgtm

and, given @nader-ziada's [comment](https://github.com/kubernetes-sigs/cluster-api-provider-azure/pull/1332#issuecomment-857712187) above

/approve 

/pony

**Comment by devigned** [comment]
> this fixes #819 and #1361, correct?

#1361, yes. But #819, no, not until cordon and drain are added to the AzureMachinePoolMachine delete.
